cluster/engine.go

```diff
@@ -741,31 +727,59 @@ func (e *Engine) refreshVolume(IDOrName string) error {
 // true, each container will be inspected.
 // FIXME: unexport this method after mesos scheduler stops using it directly
 func (e *Engine) RefreshContainers(full bool) error {
-	e.Lock()
-	defer e.Unlock()
 	opts := types.ContainerListOptions{
 		All:  true,
 		Size: false,
 	}
+	// Get the current list of containers and cache it
+	e.RLock()
+	currentContainerIDs := make([]string, 0, len(e.containers))
+	for k := range e.containers {
+		currentContainerIDs = append(currentContainerIDs, k)
+	}
+	e.RUnlock()
+
 	ctx, cancel := context.WithTimeout(context.TODO(), requestTimeout)
 	defer cancel()
 	containers, err := e.apiClient.ContainerList(ctx, opts)
-	e.checkConnectionErr(err, true)
+	e.CheckConnectionErr(err)
 	if err != nil {
 		return err
 	}
 
+	// Container IDs missing from ContainerList but are in currentContainerIDs should be delete from e.containers
+	removeContainer := func(s []string, id string) []string {
+		for i, cachedID := range s {
+			if cachedID == id {
+				return append(s[:i], s[i+1:]...)
+			}
+		}
+		return s
+	}
+	// Remove valid containers, whatever left in currentContainerIDs should be deleted from e.containers
+	for _, c := range containers {
+		currentContainerIDs = removeContainer(currentContainerIDs, c.ID)
```

Comment:
It would be a little easier to read (and slightly more efficient) if you created a new array called `missingContainerIDs` or something. Then just iterate through `currentContainerIDs`, and for each ID, if it's not in `containers`, add it to `missingContainerIDs`.

Reply:
After what I have witnessed in large deployments, I totally agree, performance should be taking with extra care. I compared 3 different approaches. 
- "Slicing" as I initially implemented 
- "nested iteration" 
- "building a map out of slice and query it"

I forced the worst case scenario on all 3 and here is what I got:
https://gist.github.com/dani-docker/897271ce9dead777ac518294cb03d82d

nestedIteration took 1.263614705s 
mapCheck took 3.755131ms 
slicing took 787.9808ms 

Are you ok with the map approach?

Reply:
Yeah, sounds fine. Thank you for doing a benchmark.